src/sentry/eventstore/reprocessing/redis.py

```diff
@@ -0,0 +1,166 @@
+        # belonging to a single group.
+        event_count = 0
+        for primary_hash in old_primary_hashes:
+            key = _get_old_primary_hash_subset_key(project_id, group_id, primary_hash)
+            event_count += self.redis.llen(key)
+        return event_count
+
+    def pop_batched_events(
+        self, project_id: int, group_id: int, primary_hash: str
+    ) -> tuple[list[str], datetime | None, datetime | None]:
+        """
+        For redis key pointing to a list of buffered events structured like
+        `event id;datetime of event`, returns a list of event IDs, the
+        earliest datetime, and the latest datetime.
+        """
+        event_ids_batch = []
+        min_datetime = None
+        max_datetime = None
+        key = _get_old_primary_hash_subset_key(project_id, group_id, primary_hash)
+
+        for row in self.redis.lrange(key, 0, -1):
+            datetime_raw, event_id = row.split(";")
+            datetime = to_datetime(float(datetime_raw))
+
+            assert datetime is not None
+
+            if min_datetime is None or datetime < min_datetime:
+                min_datetime = datetime
+            if max_datetime is None or datetime > max_datetime:
+                max_datetime = datetime
+
+            event_ids_batch.append(event_id)
+
+        self.redis.delete(key)
+
+        return event_ids_batch, min_datetime, max_datetime
+
+    def get_old_primary_hashes(self, project_id: int, group_id: int) -> set[Any]:
+        # This is a meta key that contains old primary hashes. These hashes are then
+        # combined with other values to construct a key that points to a list of
+        # tombstonable events.
+        primary_hash_set_key = f"re2:tombstone-primary-hashes:{project_id}:{group_id}"
+
+        return self.redis.smembers(primary_hash_set_key)
+
+    def expire_hash(
+        self,
+        project_id: int,
+        group_id: int,
+        event_id: str,
+        date_val: datetime,
+        old_primary_hash: str,
+    ) -> None:
+        event_key = _get_old_primary_hash_subset_key(project_id, group_id, old_primary_hash)
+        self.redis.lpush(event_key, f"{to_timestamp(date_val)};{event_id}")
+        self.redis.expire(event_key, settings.SENTRY_REPROCESSING_TOMBSTONES_TTL)
+
+    def add_hash(self, project_id: int, group_id: int, hash: str) -> None:
+        primary_hash_set_key = f"re2:tombstone-primary-hashes:{project_id}:{group_id}"
+
+        self.redis.sadd(primary_hash_set_key, hash)
+        self.redis.expire(primary_hash_set_key, settings.SENTRY_REPROCESSING_TOMBSTONES_TTL)
+
+    def get_remaining_event_count(
+        self, project_id: int, old_group_id: int, datetime_to_event: list[tuple[datetime, str]]
+    ) -> int:
+        # We explicitly cluster by only project_id and group_id here such that our
+        # RENAME command later succeeds.
+        key = f"re2:remaining:{{{project_id}:{old_group_id}}}"
```

Comment:
Can we move this key generation into a function? It's really easy to make a mistake right now